Add "selectable exceptions" to `$&catch`

jpco opened the pull request:
This PR adds the "selectable exceptions" behavior proposed in #165.

With this PR, the following two are equivalent:

```
catch @ e type rest {
    if {!~ $e error} {
        throw $e $type $rest
    }
    # handle the error
} {
    # do something risky
}
```
```
catch error @ from rest {
    # handle it
} {
    # do something risky
}
```
Note that when an exception is "selected", the first term isn't passed to the catcher.  This is a little more magic than necessary, but it does make the code read very nice.

This version of "selectable exceptions" is fairly conservative; your only options are `catch foo @ {catcher} {body}` or `catch @ {catcher} {body}`.  I think this is fine; it's simple in simple cases and the complicated version isn't actually all that complicated.  If people want more sophisticated catch-like mechanisms, they can write them; the default shell shouldn't have much magic.  (Admittedly, this is a more-than-minimal amount of magic, but in practical use, I think the "ROI" is good.  Better than `retry`, for sure.)

As usual, there are alternatives to this PR, all of which I'm pretty open to.  I'd love to rip out `retry`, ever I saw since memreflect suggest it :)

jpco committed:
Move 'retry' exception handling from $&catch to fn-catch

With this change, the $&catch primitive is about as complex as a
primitive should be.  Arguably, 'retry' should be removed from the shell
entirely, but this is a backwards-compatible half-measure, and
it does demonstrate how to "reimplement" retry in case it is actually
removed and people want to add it back.

jpco committed:
Add "selectable exceptions" for catch

This allows users to say `catch return @ value {echo $value} {command}`
and have that `catch` only catch return, and "pass through" anything
else.

jpco committed:
Add selectable exceptions to man page

memreflect commented:
This is good work!
I fully agree that for a shell like _es_, there are compromises that must be made, and simplifying things to just `catch [exception] catcher body` is a good one.

There is a [regression in `$&catch`](https://wryun.github.io/es-shell/mail-archive/msg01020.html): if a catcher throws an exception while signals are blocked, they remain blocked:

```
; forever {}
^C
; $&catch @ { throw hello } { throw world }
uncaught exception: hello
; forever {}
^C^C^C^C^C^C^C^C^C^C    # can't interrupt loop
```

Restoring the original `$&catch` implementation and simply dropping `retry` behavior fixes things.

jpco committed:
Make sure signals are unblocked after exceptions.

Also add a test case for signal blocking/unblocking in exception
handlers.

jpco committed:
Undo the $&catch to fn-catch move for `retry` handling

That move adds a lot more complexity than it removes, so we'll just
avoid changing $&catch for now and instead add selectable exceptions to
fn-catch.

jpco commented:
Well, that's disappointing, I was feeling good about having removed that nested `CatchException` in `$&catch`.

I fixed it -- and added a test case that I think should prevent fools like me from breaking it again -- but now I'm not so sure that moving the `retry` handling from `$&catch` to initial.es is worthwhile.  It adds a lot to the implementation of `fn-catch` and doesn't remove much of anything from `$&catch`.  Undid it.

For reasons I don't totally understand, I had to add a `SIGCHK()` in `$&catch` for the tests to work.  It makes some sense that we should make sure to actually throw any signal that came in when we leave the catcher.

memreflect commented:
> Well, that's disappointing, I was feeling good about having removed that nested `CatchException` in `$&catch`.

Understandable&mdash;i encountered the same issue when i was attempting to adapt Scheme's `guard` procedure to _es_ in response to the originally proposed idea.

> For reasons I don't totally understand, I had to add a `SIGCHK()` in `$&catch` for the tests to work. It makes some sense that we should make sure to actually throw any signal that came in when we leave the catcher.

Inside a catcher, signals are blocked, so when `kill -INT $pid` fires the `SIGINT` signal, it is caught for later transformation into an exception.
This transformation occurs by executing `SIGCHK()`, and that doesn't happen in the REPL until after you exit the `catch`, so your `$thrown` variable is never set to the correct value.

I can't quite understand why throwing `signal` exceptions while executing a catcher is undesirable in the first place.
The idea is probably that an incoming Unix signal shouldn't interrupt a catcher currently handling a `signal` exception, like how things work in C with the non-realtime signals.
On the other hand, dropping `blocksignals()` and `unblocksignals()` in `$&catch` means `$&catch @ e {forever {}} {throw x}` won't require you to use `kill -KILL $pid` on the shell just to terminate things.

jpco commented:
> I can't quite understand why throwing signal exceptions while executing a catcher is undesirable in the first place.
The idea is probably that an incoming Unix signal shouldn't interrupt a catcher currently handling a signal exception, like how things work in C with the non-realtime signals.

I think this is exactly it, and I think I lean towards it being the most reasonable (that is, least surprising) behavior.

Hypothetically, we could expose it to users better if we added blocked state to `signals`; maybe `$&catch` would set `signals = blocked $signals` and users could set `signals = $signals(2 ...)` if they wanted to unblock signal delivery during a particularly fancy exception handler.

I've been thinking a little about how handling "all signals" is a thing that's hard to wrangle in _es_ -- perhaps a `$&signals` primitive that will return all the signals the shell knows about would be useful too.

jpco committed:
Merge branch 'master' into catchier

jpco committed:
Merge remote-tracking branch 'upstream/master' into catchier

jpco committed:
Put "selectable exceptions" directly in $&catch

jpco committed:
Use selectable exceptions for a couple more things
